File: backend/src/services/chat_stream_service.py
"""
Chat Stream 服务层
聊天流相关业务逻辑处理
"""

from typing import Optional
from models.chat_stream import (
    ChatStream, ChatStreamCreate, ChatStreamUpdate, 
    ChatStreamQuery, ChatStreamListData
)
from models.database import OrderDirection
from core.database_manager import database_manager
import logging

logger = logging.getLogger("HMML")


class ChatStreamService:
    """Chat Stream 服务类"""
    
    def __init__(self):
        self.table_name = "chat_streams"
    
    async def get_chat_stream_list(self, query: ChatStreamQuery) -> ChatStreamListData:
        """
        获取聊天流列表（分页）
        
        Args:
            query: 查询参数
            
        Returns:
            聊天流列表数据
        """
        try:
            logger.debug(f"查询聊天流列表，参数: {query}")
            
            # 获取数据库操作器
            operator = database_manager.get_maibot_operator()
            if not operator:
                raise RuntimeError("数据库连接不可用")
            
            # 构建WHERE条件
            where_conditions = {}
            
            if query.stream_id:
                where_conditions["stream_id"] = f"%{query.stream_id}%"
            
            if query.group_platform:
                where_conditions["group_platform"] = query.group_platform
            
            if query.group_id:
                where_conditions["group_id"] = f"%{query.group_id}%"
            
            if query.group_name:
                where_conditions["group_name"] = f"%{query.group_name}%"
            
            if query.platform:
                where_conditions["platform"] = query.platform
            
            if query.user_platform:
                where_conditions["user_platform"] = query.user_platform
            
            if query.user_id:
                where_conditions["user_id"] = f"%{query.user_id}%"
            
            if query.user_nickname:
                where_conditions["user_nickname"] = f"%{query.user_nickname}%"
            
            if query.user_cardname:
                where_conditions["user_cardname"] = f"%{query.user_cardname}%"
            
            # 使用分页查询
            result = operator.find_with_pagination(
                table_name=self.table_name,
                page=query.page,
                page_size=query.pageSize,
                where_conditions=where_conditions,
                order_by="id",
                order_dir=OrderDirection.DESC
            )
            
            # 转换为模型对象
            items = []
            for row in result.items:
                # 处理可能为 None 的字段
                for field in ['group_platform', 'group_id', 'group_name', 'user_platform', 
                             'user_nickname', 'user_cardname']:
                    if row.get(field) is None:
                        row[field] = ""
                
                items.append(ChatStream(**row))
            
            logger.info(f"查询聊天流列表成功，共 {len(items)} 条记录")
            
            return ChatStreamListData(
                items=items,
                totalPages=result.total_pages,
                currentPage=result.current_page,
                pageSize=result.page_size
            )
            
        except Exception as error:
            logger.error(f"查询聊天流列表失败: {error}")
            raise
    
    async def create_chat_stream(self, chat_stream: ChatStreamCreate) -> int:
        """
        创建聊天流
        
        Args:
            chat_stream: 聊天流数据
            
        Returns:
            新创建记录的ID
        """
        try:
            logger.debug(f"创建聊天流: {chat_stream.stream_id}")
            
            # 获取数据库操作器
            operator = database_manager.get_maibot_operator()
            if not operator:
                raise RuntimeError("数据库连接不可用")
            
            # 准备插入数据
            insert_data = {
                'stream_id': chat_stream.stream_id,
                'create_time': chat_stream.create_time,
                'group_platform': chat_stream.group_platform,
                'group_id': chat_stream.group_id,
                'group_name': chat_stream.group_name,
                'last_active_time': chat_stream.last_active_time,
                'platform': chat_stream.platform,
                'user_platform': chat_stream.user_platform,
                'user_id': chat_stream.user_id,
                'user_nickname': chat_stream.user_nickname,
                'user_cardname': chat_stream.user_cardname
            }
            
            # 执行插入
            result = operator.insert(self.table_name, insert_data)
            
            if result.success and result.last_insert_id:
                logger.info(f"创建聊天流成功，ID: {result.last_insert_id}")
                return result.last_insert_id
            else:
                raise RuntimeError("插入操作失败")
            
        except Exception as error:
            logger.error(f"创建聊天流失败: {error}")
            raise
    
    async def update_chat_stream(self, chat_stream: ChatStreamUpdate) -> bool:
        """
        更新聊天流
        
        Args:
            chat_stream: 聊天流数据
            
        Returns:
            是否更新成功
        """
        try:
            logger.debug(f"更新聊天流，ID: {chat_stream.id}")
            
            # 获取数据库操作器
            operator = database_manager.get_maibot_operator()
            if not operator:
                raise RuntimeError("数据库连接不可用")
            
            # 准备更新数据
            update_data = {
                'stream_id': chat_stream.stream_id,
                'create_time': chat_stream.create_time,
                'group_platform': chat_stream.group_platform,
                'group_id': chat_stream.group_id,
                'group_name': chat_stream.group_name,
                'last_active_time': chat_stream.last_active_time,
                'platform': chat_stream.platform,
                'user_platform': chat_stream.user_platform,
                'user_id': chat_stream.user_id,
                'user_nickname': chat_stream.user_nickname,
                'user_cardname': chat_stream.user_cardname
            }
            
            # WHERE条件
            where_conditions = {'id': chat_stream.id}
            
            # 执行更新
            result = operator.update(self.table_name, update_data, where_conditions)
            
            if result.success:
                logger.info(f"更新聊天流成功，ID: {chat_stream.id}")
                return True
            else:
                return False
            
        except Exception as error:
            logger.error(f"更新聊天流失败: {error}")
            raise
    
    async def delete_chat_stream(self, chat_stream_id: int) -> bool:
        """
        删除聊天流
        
        Args:
            chat_stream_id: 聊天流ID
            
        Returns:
            是否删除成功
        """
        try:
            logger.debug(f"删除聊天流，ID: {chat_stream_id}")
            
            # 获取数据库操作器
            operator = database_manager.get_maibot_operator()
            if not operator:
                raise RuntimeError("数据库连接不可用")
            
            # WHERE条件
            where_conditions = {'id': chat_stream_id}
            
            # 执行删除
            result = operator.delete(self.table_name, where_conditions)
            
            if result.success and result.affected_rows > 0:
                logger.info(f"删除聊天流成功，ID: {chat_stream_id}")
                return True
            else:
                return False
            
        except Exception as error:
            logger.error(f"删除聊天流失败: {error}")
            raise
    
    async def get_chat_stream_by_id(self, chat_stream_id: int) -> Optional[ChatStream]:
        """
        根据ID获取聊天流
        
        Args:
            chat_stream_id: 聊天流ID
            
        Returns:
            聊天流对象或None
        """
        try:
            logger.debug(f"根据ID查询聊天流: {chat_stream_id}")
            
            # 获取数据库操作器
            operator = database_manager.get_maibot_operator()
            if not operator:
                raise RuntimeError("数据库连接不可用")
            
            # WHERE条件
            where_conditions = {'id': chat_stream_id}
            
            # 执行查询
            result = operator.find_one(self.table_name, where_conditions)
            
            if result:
                # 处理可能为 None 的字段
                for field in ['group_platform', 'group_id', 'group_name', 'user_platform', 
                             'user_nickname', 'user_cardname']:
                    if result.get(field) is None:
                        result[field] = ""
                
                logger.info(f"查询聊天流成功，ID: {chat_stream_id}")
                return ChatStream(**result)
            else:
                logger.info(f"未找到聊天流，ID: {chat_stream_id}")
                return None
            
        except Exception as error:
            logger.error(f"查询聊天流失败: {error}")
            raise


# 全局服务实例
chat_stream_service = ChatStreamService()

File: backend/old/src/services/chatStreamService.ts
/**
 * ChatStream表专用操作服务
 * 提供对chat_streams表的高级操作功能
 */

import { logger } from '../core/logger';
import { databaseManager } from '../core/database';
import { pathCacheManager } from '../core/pathCacheManager';
import { 
  ChatStreamRecord, 
  ChatStreamInsertData, 
  ChatStreamUpdateData,
  ChatStreamQueryFilter,
  ChatStreamPaginationParams,
  ChatStreamQueryResponse
} from '../types/chatStream';
import { QueryFilter, PaginationParams, SortParams } from '../types/database';
import { ValidationError, NotFoundError, InternalServerError } from '../middleware/errorHandler';
import path from 'path';

export class ChatStreamService {
  private static readonly TABLE_NAME = 'chat_streams';
  private static readonly CONNECTION_NAME = 'maibot';

  /**
   * 初始化麦麦数据库连接
   */
  private static async initConnection(): Promise<void> {
    const mainRoot = pathCacheManager.getMainRoot();
    if (!mainRoot) {
      throw new NotFoundError('麦麦主程序根目录未设置，请先设置根目录缓存');
    }

    const dbPath = path.join(mainRoot, 'data', 'MaiBot.db');
    
    // 检查连接是否已存在
    if (!databaseManager.hasConnection(this.CONNECTION_NAME)) {
      if (!databaseManager.isManagerInitialized()) {
        await databaseManager.initialize();
      }
      
      await databaseManager.addDatabase(this.CONNECTION_NAME, {
        path: dbPath,
        readonly: false,
        timeout: 5000,
        fileMustExist: true
      });
      
      logger.info('麦麦数据库连接已建立');
    }
  }

  /**
   * 验证聊天流数据
   */
  private static validateChatStreamData(data: ChatStreamInsertData | ChatStreamUpdateData): void {
    const requiredFields = ['stream_id', 'group_platform', 'group_id', 'group_name', 'platform', 'user_platform', 'user_id', 'user_nickname', 'user_cardname'];
    
    for (const field of requiredFields) {
      if (!(field in data) || !data[field as keyof typeof data]) {
        throw new ValidationError(`字段 ${field} 不能为空`);
      }
    }

    // 验证数字字段
    const numericFields = ['create_time', 'last_active_time'];
    for (const field of numericFields) {
      if (field in data && typeof data[field as keyof typeof data] !== 'number') {
        throw new ValidationError(`字段 ${field} 必须是数字类型`);
      }
    }

    // 验证时间字段的合理性
    if ('create_time' in data && data.create_time !== undefined && data.create_time <= 0) {
      throw new ValidationError('create_time 必须是大于0的时间戳');
    }

    if ('last_active_time' in data && data.last_active_time !== undefined && data.last_active_time <= 0) {
      throw new ValidationError('last_active_time 必须是大于0的时间戳');
    }
  }

  /**
   * 构建查询过滤条件
   */
  private static buildQueryFilters(filter: ChatStreamQueryFilter): QueryFilter[] {
    const conditions: QueryFilter[] = [];

    if (filter.stream_id) {
      conditions.push({ field: 'stream_id', operator: '=', value: filter.stream_id });
    }

    if (filter.group_platform) {
      conditions.push({ field: 'group_platform', operator: '=', value: filter.group_platform });
    }

    if (filter.group_id) {
      conditions.push({ field: 'group_id', operator: '=', value: filter.group_id });
    }

    if (filter.group_name) {
      conditions.push({ field: 'group_name', operator: 'LIKE', value: `%${filter.group_name}%` });
    }

    if (filter.platform) {
      conditions.push({ field: 'platform', operator: '=', value: filter.platform });
    }

    if (filter.user_platform) {
      conditions.push({ field: 'user_platform', operator: '=', value: filter.user_platform });
    }

    if (filter.user_id) {
      conditions.push({ field: 'user_id', operator: '=', value: filter.user_id });
    }

    if (filter.user_nickname) {
      conditions.push({ field: 'user_nickname', operator: 'LIKE', value: `%${filter.user_nickname}%` });
    }

    if (filter.user_cardname) {
      conditions.push({ field: 'user_cardname', operator: 'LIKE', value: `%${filter.user_cardname}%` });
    }

    return conditions;
  }

  /**
   * 分页查询聊天流
   */
  public static async getChatStreams(params: ChatStreamPaginationParams): Promise<ChatStreamQueryResponse> {
    try {
      await this.initConnection();
      const operator = databaseManager.getOperator(this.CONNECTION_NAME);

      // 构建查询条件
      const conditions = params.filter ? this.buildQueryFilters(params.filter) : [];

      // 构建分页参数
      const paginationParams: PaginationParams = {
        page: params.page,
        pageSize: params.pageSize
      };

      // 构建排序参数
      const orderByParams: SortParams[] | undefined = params.orderBy ? [{
        field: params.orderBy,
        order: params.orderDir || 'ASC'
      }] : undefined;

      const result = await operator.findWithPagination<ChatStreamRecord>(
        this.TABLE_NAME,
        paginationParams,
        {
          where: conditions.length > 0 ? conditions : undefined,
          orderBy: orderByParams
        }
      );

      return {
        items: result.data,
        totalPages: result.pagination.totalPages,
        currentPage: result.pagination.page,
        pageSize: result.pagination.pageSize,
        total: result.pagination.total,
        hasNext: result.pagination.hasNext,
        hasPrev: result.pagination.hasPrev
      };

    } catch (error) {
      logger.error('查询聊天流失败:', error);
      throw error instanceof ValidationError 
        ? error 
        : new InternalServerError('查询聊天流失败');
    }
  }

  /**
   * 根据ID获取单个聊天流
   */
  public static async getChatStreamById(id: number): Promise<ChatStreamRecord | null> {
    try {
      if (!id || id < 1) {
        throw new ValidationError('ID必须是大于0的数字');
      }

      await this.initConnection();
      const operator = databaseManager.getOperator(this.CONNECTION_NAME);

      const result = await operator.findOne<ChatStreamRecord>(this.TABLE_NAME, {
        where: [{ field: 'id', operator: '=', value: id }]
      });

      return result;

    } catch (error) {
      logger.error('根据ID查询聊天流失败:', error);
      throw error instanceof ValidationError 
        ? error 
        : new InternalServerError('查询聊天流失败');
    }
  }

  /**
   * 插入新的聊天流
   */
  public static async insertChatStream(data: ChatStreamInsertData): Promise<number> {
    try {
      this.validateChatStreamData(data);
      
      await this.initConnection();
      const operator = databaseManager.getOperator(this.CONNECTION_NAME);

      const result = await operator.insertOne(this.TABLE_NAME, data);
      
      logger.info(`成功插入聊天流，ID: ${result.lastInsertId}, stream_id: ${data.stream_id}`);
      return result.lastInsertId || 0;

    } catch (error) {
      logger.error('插入聊天流失败:', error);
      throw error instanceof ValidationError 
        ? error 
        : new InternalServerError('插入聊天流失败');
    }
  }

  /**
   * 更新聊天流
   */
  public static async updateChatStream(data: ChatStreamUpdateData): Promise<void> {
    try {
      if (!data.id || data.id < 1) {
        throw new ValidationError('ID必须是大于0的数字');
      }

      this.validateChatStreamData(data);
      
      await this.initConnection();
      const operator = databaseManager.getOperator(this.CONNECTION_NAME);

      // 检查记录是否存在
      const existing = await this.getChatStreamById(data.id);
      if (!existing) {
        throw new NotFoundError(`未找到ID为 ${data.id} 的聊天流记录`);
      }

      const { id, ...updateData } = data;
      const result = await operator.update(
        this.TABLE_NAME,
        updateData,
        { where: [{ field: 'id', operator: '=', value: id }] }
      );

      if (result.affectedRows === 0) {
        throw new NotFoundError(`未找到ID为 ${id} 的聊天流记录`);
      }

      logger.info(`成功更新聊天流，ID: ${id}`);

    } catch (error) {
      logger.error('更新聊天流失败:', error);
      throw error instanceof ValidationError || error instanceof NotFoundError
        ? error 
        : new InternalServerError('更新聊天流失败');
    }
  }

  /**
   * 删除聊天流
   */
  public static async deleteChatStream(id: number): Promise<void> {
    try {
      if (!id || id < 1) {
        throw new ValidationError('ID必须是大于0的数字');
      }

      await this.initConnection();
      const operator = databaseManager.getOperator(this.CONNECTION_NAME);

      // 检查记录是否存在
      const existing = await this.getChatStreamById(id);
      if (!existing) {
        throw new NotFoundError(`未找到ID为 ${id} 的聊天流记录`);
      }

      const result = await operator.delete(this.TABLE_NAME, {
        where: [{ field: 'id', operator: '=', value: id }]
      });

      if (result.affectedRows === 0) {
        throw new NotFoundError(`未找到ID为 ${id} 的聊天流记录`);
      }

      logger.info(`成功删除聊天流，ID: ${id}`);

    } catch (error) {
      logger.error('删除聊天流失败:', error);
      throw error instanceof ValidationError || error instanceof NotFoundError
        ? error 
        : new InternalServerError('删除聊天流失败');
    }
  }

  /**
   * 根据stream_id查询聊天流
   */
  public static async getChatStreamByStreamId(streamId: string): Promise<ChatStreamRecord | null> {
    try {
      if (!streamId) {
        throw new ValidationError('stream_id不能为空');
      }

      await this.initConnection();
      const operator = databaseManager.getOperator(this.CONNECTION_NAME);

      const result = await operator.findOne<ChatStreamRecord>(this.TABLE_NAME, {
        where: [{ field: 'stream_id', operator: '=', value: streamId }]
      });

      return result;

    } catch (error) {
      logger.error('根据stream_id查询聊天流失败:', error);
      throw error instanceof ValidationError 
        ? error 
        : new InternalServerError('查询聊天流失败');
    }
  }

  /**
   * 获取聊天流统计信息
   */
  public static async getChatStreamStats(): Promise<any> {
    try {
      await this.initConnection();
      const operator = databaseManager.getOperator(this.CONNECTION_NAME);

      // 总数统计
      const totalCount = await operator.count(this.TABLE_NAME);

      // 按平台统计
      const platformStats = await operator.executeQuery(`
        SELECT platform, COUNT(*) as count 
        FROM ${this.TABLE_NAME} 
        GROUP BY platform
      `);

      // 按群组平台统计
      const groupPlatformStats = await operator.executeQuery(`
        SELECT group_platform, COUNT(*) as count 
        FROM ${this.TABLE_NAME} 
        GROUP BY group_platform
      `);

      // 最近活跃的聊天流
      const recentActiveStreams = await operator.executeQuery(`
        SELECT COUNT(*) as count 
        FROM ${this.TABLE_NAME} 
        WHERE last_active_time > ?
      `, [Date.now() - 24 * 60 * 60 * 1000]); // 24小时内

      return {
        totalCount,
        platformStats,
        groupPlatformStats,
        recentActiveCount: recentActiveStreams[0]?.count || 0
      };

    } catch (error) {
      logger.error('获取聊天流统计信息失败:', error);
      throw new InternalServerError('获取聊天流统计信息失败');
    }
  }
}
